### PR TITLE
Support ForwardRef type of work in TestRenderer

### PR DESCRIPTION
**Pull request by jquense**

### CODE REVIEW

**Comment by jquense** [comment]
Not sure how to update ShallowRenderer either

**Comment by jquense** [comment]
i don't understand why the react CI hates me so much

File: packages/react-test-renderer/src/ReactShallowRenderer.js [hunk 1]
```diff
@@ -96,7 +97,9 @@ class ReactShallowRenderer {
     if (this._instance) {
       this._updateClassComponent(element, this._context);
     } else {
-      if (shouldConstruct(element.type)) {
+      if (isForwardRef(element)) {
+        this._rendered = element.type.render(element.props, null);
```
Comment (jquense) [inline packages/react-test-renderer/src/ReactShallowRenderer.js]: This seems fine in the case where you're rendering just host elements, but maybe not helpful for the HOC case where you wrap and render a single composite component...

**Comment by bvaughn** [comment]
Circle CI is down right now for ReactJS.org repo too:
> #!/bin/bash -eo pipefail
yarn install
/bin/bash: yarn: command not found
Exited with code 127

**Comment by bvaughn** [comment]
Looks like CI's failing because of nodejs/docker-node/issues/651

File: packages/react-test-renderer/package.json [hunk 2]
```diff
@@ -17,16 +13,11 @@
   "dependencies": {
     "fbjs": "^0.8.16",
     "object-assign": "^4.1.1",
-    "prop-types": "^15.6.0"
+    "prop-types": "^15.6.0",
+    "react-is": "^16.3.0-alpha.2"
```
Comment (bvaughn) [inline packages/react-test-renderer/package.json]: Will this require updates to our release/build script to keep these deps in sync? Needs follow up.

File: packages/react-test-renderer/src/__tests__/ReactShallowRenderer-test.js [hunk 3]
```diff
@@ -168,6 +168,26 @@ describe('ReactShallowRenderer', () => {
...
+    const shallowRenderer = createRenderer();
+    const result = shallowRenderer.render(<SomeComponent />);
+
+    expect(result.type).toBe('div');
+    expect(result.props.children).toEqual([
+      null,
+      <span className="child1" />,
+      <span className="child2" />,
+    ]);
+  });
```
Comment (bvaughn) [inline packages/react-test-renderer/src/__tests__/ReactShallowRenderer-test.js]: Suggestion:
```js
it('should handle ForwardRef', () => {
  const testRef = React.createRef();
  const SomeComponent = React.forwardRef((props, ref) => {
    expect(ref).toEqual(testRef);
    return (
      <div>
        <span className="child1" />
        <span className="child2" />
      </div>
    )
  });

  const shallowRenderer = createRenderer();
  const result = shallowRenderer.render(<SomeComponent ref={testRef} />);

  expect(result.type).toBe('div');
  expect(result.props.children).toEqual([
    <span className="child1" />,
    <span className="child2" />,
  ]);
});
```

File: packages/react-test-renderer/src/ReactShallowRenderer.js (hunk 1 above)
Comment (bvaughn) [reply packages/react-test-renderer/src/ReactShallowRenderer.js]: I think this line should be:
```js
this._rendered = element.type.render(element.props, element.ref);
```

I'm not sure I understand your thought though?

**Comment by bvaughn** [comment]
Thanks for this PR, by the way! 😁

File: packages/react-is/src/ReactIs.js [hunk 4]
```diff
@@ -34,7 +34,7 @@ export function typeOf(object: any) {
           case REACT_STRICT_MODE_TYPE:
             return type;
           default:
-            const $$typeofType = type.$$typeof;
+            const $$typeofType = type && type.$$typeof;
```
[the remark this answers is not in the document]
Comment (jquense) [reply packages/react-is/src/ReactIs.js]: probably not in practice but the ShallowRenderer tests do `<NullIdentifer>` which was how i was hitting this.

File: packages/react-test-renderer/package.json (hunk 2 above)
Comment (jquense) [reply packages/react-test-renderer/package.json]: hmm i don't know, I was conflating workspaces and lerna, the latter of which handles the updates

File: packages/react-test-renderer/src/ReactShallowRenderer.js (hunk 1 above)
Comment (jquense) [reply packages/react-test-renderer/src/ReactShallowRenderer.js]: I was thinking of this case:

```js
function withFoo(Component) {
  const Foo = (props) => <Component ref={props.forwardedRef} />
  return React.forwardRef((props, ref) => <Foo forwardedRef={ref} />)
}
shallow(<WithFooComponent><div/></WithFooComponent>)
```

You'd get a different result than without the `forwardRef` but I guess that's true of any HoC thing, might even help hide the HoC guts a bit more with is nice.

File: packages/react-is/src/ReactIs.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (bvaughn) [reply packages/react-is/src/ReactIs.js]: Oh I see. 👍

File: packages/react-test-renderer/package.json (hunk 2 above)
Comment (bvaughn) [reply packages/react-test-renderer/package.json]: These versions are bumped by our release script manually. It's fine. I'll add it to my list of follow up things.

File: packages/react-test-renderer/src/ReactShallowRenderer.js (hunk 1 above)
Comment (bvaughn) [reply packages/react-test-renderer/src/ReactShallowRenderer.js]: I see. In general, I'm not sure how valuable shallow testing with anything refs or HOC related is. 😄

File: packages/react-test-renderer/package.json (hunk 2 above)
Comment (bvaughn) [reply packages/react-test-renderer/package.json]: Addressed via 610df95

**Comment by bvaughn** [comment]
Lint, Flow, and tests pass for me locally so I'm going to move forward in spite of the CI/Docker issue.

**Pull request merged**